arch/xtensa/core/mpu.c

```diff
@@ -701,3 +707,419 @@ void xtensa_mpu_init(void)
+		break;
+	case XTENSA_MPU_ACCESS_P_RO_U_RO:
+	case XTENSA_MPU_ACCESS_P_RO_U_NA:
+		*readable = true;
+		*writable = false;
+		*executable = false;
+		break;
+		break;
+	case XTENSA_MPU_ACCESS_P_RX_U_NA:
+	case XTENSA_MPU_ACCESS_P_RX_U_RX:
+		*readable = true;
+		*writable = false;
+		*executable = true;
+		break;
+	case XTENSA_MPU_ACCESS_P_RW_U_NA:
+	case XTENSA_MPU_ACCESS_P_RW_U_RW:
+	case XTENSA_MPU_ACCESS_P_RW_U_RWX:
+	case XTENSA_MPU_ACCESS_P_RW_U_RO:
+		*readable = true;
+		*writable = true;
+		*executable = false;
+		break;
+	case XTENSA_MPU_ACCESS_P_RWX_U_NA:
+	case XTENSA_MPU_ACCESS_P_RWX_U_RX:
+	case XTENSA_MPU_ACCESS_P_RWX_U_RWX:
+		*readable = true;
+		*writable = true;
+		*executable = true;
+		break;
+	case XTENSA_MPU_ACCESS_P_WO_U_WO:
+		*readable = false;
+		*writable = true;
+		*executable = false;
+		break;
+	case XTENSA_MPU_ACCESS_P_NA_U_NA:
+	case XTENSA_MPU_ACCESS_P_NA_U_X:
+	default:
+		*readable = false;
+		*writable = false;
+		*executable = false;
+		break;
+	}
+}
+
+int arch_mem_domain_init(struct k_mem_domain *domain)
+{
+	domain->arch.mpu_map = xtensa_mpu_map_fg_kernel;
+
+	return 0;
+}
+
+int arch_mem_domain_max_partitions_get(void)
+{
+	/*
+	 * Due to each memory region requiring 2 MPU entries to describe,
+	 * it is hard to figure out how many partitions are available.
+	 * For example, if all those partitions are contiguous, it only
+	 * needs 2 entries (1 if the end of region already has an entry).
+	 * If they are all disjoint, it will need (2 * n) entries to
+	 * describe all of them. So just use CONFIG_MAX_DOMAIN_PARTITIONS
+	 * here and let the application set this instead.
+	 */
+	return CONFIG_MAX_DOMAIN_PARTITIONS;
+}
+
+int arch_mem_domain_partition_remove(struct k_mem_domain *domain,
+				     uint32_t partition_id)
+{
+	int ret;
+	uint32_t perm;
+	bool readable, writable, executable;
+	struct xtensa_mpu_map *map = &domain->arch.mpu_map;
+	struct k_mem_partition *partition = &domain->partitions[partition_id];
+	uintptr_t end_addr = partition->start + partition->size;
+
+	if (end_addr <= partition->start) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/*
+	 * This is simply to get rid of the user permissions and retain
+	 * whatever the kernel permissions are. So that we won't be
+	 * setting the memory region permission incorrectly, for example,
+	 * marking read only region writable.
+	 *
+	 * Note that Zephyr does not do RWX partitions so we can treat it
+	 * as invalid.
+	 */
+	decode_kernel_permission(partition->attr, &readable, &writable, &executable);
+
+	if (writable) {
+		if (readable) {
+			perm = XTENSA_MPU_ACCESS_P_RW_U_NA;
+		} else {
+			perm = XTENSA_MPU_ACCESS_P_WO_U_WO;
+		}
+	} else if (executable) {
+		if (readable) {
+			perm = XTENSA_MPU_ACCESS_P_RX_U_NA;
+		} else {
+			perm = XTENSA_MPU_ACCESS_P_X_U_NA;
+		}
+	} else if (readable) {
+		perm = XTENSA_MPU_ACCESS_P_RO_U_NA;
+	} else {
+		/* fail safe to have no access at all. */
+		perm = XTENSA_MPU_ACCESS_P_NA_U_NA;
+	}
```

Comment:
Might be cleaner at runtime to pack the r/w/x bits into an array index and define a table with 8 entries.   More code but maybe fewer instructions?

Reply:
Converted it to switch/case.